src/gbt.cpp

```diff
@@ -85,124 +65,88 @@ bool scanGBTPhasesLocal(localArgs *la, const uint32_t ohN, const uint32_t N, con
-
-    // Write the configuration
-    writeGBTConfigLocal(&la, ohN, gbtN, config);
-
-    rtxn.abort();
-} //End writeGBTConfig(...)
-
-bool writeGBTConfigLocal(localArgs *la, const uint32_t ohN, const uint32_t gbtN, const gbt::config_t &config)
-{
     LOG4CPLUS_INFO(logger, stdsprintf("Writing the configuration of OH #%u - GBTX #%u.", ohN, gbtN));
 
+    config_t config{};
     // ohN check
-    const uint32_t ohMax = readReg(la, "GEM_AMC.GEM_SYSTEM.CONFIG.NUM_OF_OH");
+    const uint32_t ohMax = readReg("GEM_AMC.GEM_SYSTEM.CONFIG.NUM_OF_OH");
     if (ohN >= ohMax)
-        EMIT_RPC_ERROR(la->response, stdsprintf("The ohN parameter supplied (%u) exceeds the number of OH's supported by the CTP7 (%u).", ohN, ohMax), true);
+        EMIT_RPC_ERROR(stdsprintf("The ohN parameter supplied (%u) exceeds the number of OH's supported by the CTP7 (%u).", ohN, ohMax), true);
 
     // gbtN check
-    if (gbtN >= gbt::GBTS_PER_OH)
-        EMIT_RPC_ERROR(la->response, stdsprintf("The gbtN parameter supplied (%u) exceeds the number of GBT's per OH (%u).", gbtN, gbt::GBTS_PER_OH), true);
+    if (gbtN >= GBTS_PER_OH)
+        EMIT_RPC_ERROR(stdsprintf("The gbtN parameter supplied (%u) exceeds the number of GBT's per OH (%u).", gbtN, GBTS_PER_OH), true);
 
     // Write all the registers
-    for (size_t address = 0; address < gbt::CONFIG_SIZE; address++) {
-        if (writeGBTRegLocal(la, ohN, gbtN, static_cast<uint16_t>(address), config[address]))
+    for (size_t address = 0; address < CONFIG_SIZE; address++) {
+        if (writeGBTRegLocal(ohN, gbtN, static_cast<uint16_t>(address), config[address]))
             return true;
     }
 
     return false;
-} //End writeGBTConfigLocal(...)
-
-void writeGBTPhase(const RPCMsg *request, RPCMsg *response)
-{
-    GETLOCALARGS(response);
-
-    // Get the keys
-    const uint32_t ohN = request->get_word("ohN");
-    const uint32_t vfatN = request->get_word("vfatN");
-    const uint8_t phase = request->get_word("phase");
-
-    // Write the phase
-    writeGBTPhaseLocal(&la, ohN, vfatN, phase);
 
-    rtxn.abort();
-} //End writeGBTPhase
+} //End writeGBTConfig
 
-bool writeGBTPhaseLocal(localArgs *la, const uint32_t ohN, const uint32_t vfatN, const uint8_t phase)
+bool WriteGBTPhase::operator()(const uint32_t &ohN, const uint32_t &vfatN, const uint8_t &phase) const
 {
+
     LOG4CPLUS_INFO(logger, stdsprintf("Writing the VFAT #%u phase of OH #%u.", vfatN, ohN));
 
     // ohN check
-    const uint32_t ohMax = readReg(la, "GEM_AMC.GEM_SYSTEM.CONFIG.NUM_OF_OH");
+    const uint32_t ohMax = readReg("GEM_AMC.GEM_SYSTEM.CONFIG.NUM_OF_OH");
     if (ohN >= ohMax)
-        EMIT_RPC_ERROR(la->response, stdsprintf("The ohN parameter supplied (%u) exceeds the number of OH's supported by the CTP7 (%u).", ohN, ohMax), true);
+        EMIT_RPC_ERROR(stdsprintf("The ohN parameter supplied (%u) exceeds the number of OH's supported by the CTP7 (%u).", ohN, ohMax), true);
```

Comment:
For all of these checks against a min/max, let's just raise `std::range_error` with the error message as it currently is

Reply:
Since we're refactoring, would it make sense to have a set of `checkOHN(int)` functions that would read the register and perform the appropriate checks?

Reply:
In the PR which introduced the set of `check*` functions, it was decided not to have the `checkOHN(std::uint32_t)` for performance reasons... because the OH check requires a register read.

Over time, the `ohMax` value has been half-ported to a GEM station constant, but it shouldn't since it is a property of the firmware are not of the GEM station. I'm more and more thinking that a set of properties (such a the maximum number of OH supported by the firmware) will have be initialized at runtime during module initialization.

Also, I think it is worth deciding where the range checks should be done. It can obviously be done in every function, but we could also restrict the checks to exported functions (i.e. functors).

Reply:
> In the PR which introduced the set of `check*` functions, it was decided not to have the `checkOHN(std::uint32_t)` for performance reasons... because the OH check requires a register read.
> 
> Over time, the `ohMax` value has been half-ported to a GEM station constant, but it shouldn't since it is a property of the firmware are not of the GEM station. I'm more and more thinking that a set of properties (such a the maximum number of OH supported by the firmware) will have be initialized at runtime during module initialization.
> 

The only problem I'd have with this is that this could be happening many times.
I'd definitely like to think how to reduce the number of these checks, and maybe the answer is in the `Hw<Device>` class in `cmsgemos`, which can actually be stateful, and this relates to your next point.

> Also, I think it is worth deciding where the range checks should be done. It can obviously be done in every function, but we could also restrict the checks to exported functions (i.e. functors).

If we ensure that no invalid values are ever passed to the module code, then we can remove (most of) these checks entirely, which is naively what I think would be better, and the module code can really focus just on the register actions.

Now is definitely the right time to put some additional thought into the architecture, since we're making a quite big change anyway, though might be worth it to create a new issue to discuss potential changes, if there end up being a reasonably long list of them.

Reply:
> > In the PR which introduced the set of `check*` functions, it was decided not to have the `checkOHN(std::uint32_t)` for performance reasons... because the OH check requires a register read.
> > Over time, the `ohMax` value has been half-ported to a GEM station constant, but it shouldn't since it is a property of the firmware are not of the GEM station. I'm more and more thinking that a set of properties (such a the maximum number of OH supported by the firmware) will have be initialized at runtime during module initialization.
> 
> The only problem I'd have with this is that this could be happening many times.

How often do you plan to reconnect to the CTP7? I understood that it would be done once or twice per CTP7 at `cmsgemos` initialization. I wouldn't call that many times. ;-)  

> I'd definitely like to think how to reduce the number of these checks, and maybe the answer is in the `Hw<Device>` class in `cmsgemos`, which can actually be stateful, and this relates to your next point.
> 
> > Also, I think it is worth deciding where the range checks should be done. It can obviously be done in every function, but we could also restrict the checks to exported functions (i.e. functors).
> 
> If we ensure that no invalid values are ever passed to the module code, then we can remove (most of) these checks entirely, which is naively what I think would be better, and the module code can really focus just on the register actions.

If the RPC modules will only ever be called through the `Hw<Device>` classes (and I think that will be the case), I don't have strong arguments against the idea. Usually, the recommendation is to completely check the values coming from the outside, but since we have a complete control of the system....

> Now is definitely the right time to put some additional thought into the architecture, since we're making a quite big change anyway, though might be worth it to create a new issue to discuss potential changes, if there end up being a reasonably long list of them.

:+1: